Add kubectl tutorial to walk through some common developer tasks (#85)
alphagov/govuk-kubernetes-cluster-user-docs: source/get-started/tutorials/k8s-command-line/index.html.md

---
title: Kubernetes command line
weight: 20
last_reviewed_on: 2023-04-18
review_in: 6 months
---

> # Use the Kubernetes command line tooling

Reviewer: ```suggestion
# Familiarise yourself with `kubectl`, the Kubernetes command-line tool
```

> This tutorial will teach you how to use the Kubernetes command line tooling.

Reviewer: ```suggestion
This tutorial will show you how to use the `kubectl` command line tool.
```

> It requires that you have [Set up tools to use the GOV.UK Kubernetes platform](https://govuk-k8s-user-docs.publishing.service.gov.uk/get-started/set-up-tools/) and that you have successfully [tested your access](https://govuk-k8s-user-docs.publishing.service.gov.uk/get-started/access-eks-cluster/#test-your-access).

Reviewer: ```suggestion
You need to have completed the [set up instructions](/get-started/set-up-tools/) and [tested your access](/get-started/access-eks-cluster/#test-your-access).
```

> ## Fetch logs from a pod

Reviewer: This duplicates https://govuk-k8s-user-docs.publishing.service.gov.uk/manage-app/get-app-info/#view-app-logs, but with extra steps.

It might be nicer to fold this into the [deploy-monitor-logs tutorial](https://govuk-k8s-user-docs.publishing.service.gov.uk/get-started/tutorials/app-update-deploy-monitor-logs/) since that already has a "check the logs" step. It'd just be a single extra command there.

```
kubectl -n apps get pods
```

This will return a list, including output like:

```
NAME                                                              READY   STATUS             RESTARTS           AGE
account-api-568d46d6f7-4pnld                                      2/2     Running            0                  34m
account-api-568d46d6f7-txwsx                                      2/2     Running            0                  34m
account-api-dbmigrate-6n5mp                                       0/1     Completed          0                  35m
account-api-worker-54597b666c-jlcfn                               1/1     Running            0                  34m
asset-manager-7bd745d655-tzlgz                                    2/2     Running            0                  15h
...
```

As you can see above, each pod either runs the application (we can see two pods running Account API, for example) or runs a worker or cron job for the application (we can see one pod running the Account API application worker).

You can fetch the logs for all running containers in a pod. Take one of the pods from the output above:

```
kubectl -n apps logs account-api-568d46d6f7-4pnld
```

> Or you can fetch the logs using its class name. This will choose one of the Account API pods at random, much like the old jumpbox methods:

Author: "class name" probably the wrong term to use here - what should it be?

Reviewer: You could use the deployment or you could use labels. We can update [view-app-logs](https://govuk-k8s-user-docs.publishing.service.gov.uk/manage-app/get-app-info/#view-app-logs) to reflect that better.

Author: Thanks Chris. I don't see a reference to "label" in the doc you link to, so not sure how to action that.

Sounds like I should at least tweak this to the following though:

```suggestion
Or you can fetch the logs using its 'app deployment' name. This will choose one of the Account API pods at random, much like the old jumpbox methods:
```

```

This is because we set a readonly context earlier. We need write permissions to be able to open a console. Let's correct this now:

```
eval $(gds aws govuk-integration-poweruser -e --art 8h)
```

Now when you try the command to open a Rails console, it succeeds!

## Open a shell on a pod

With your 'write' permissions set, you should be able to open a shell:

```
k exec -it deploy/whitehall-admin -- bash
```

Note that the prompt will have prefix that may feel a little unfamiliar: `I have no name!@whitehall-admin-6bb47c48d-wkkxk:/app$`.

If you try to open a shell on Router, it will fail:

```
$ k exec -it deploy/router -- bash                                                               
Defaulted container "app" out of: app, nginx
error: Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "e870a2214da6273408c75773234b09c245210d3f1dff3b200c096f24c7259f7f": OCI runtime exec failed: exec failed: unable to start container process: exec: "bash": executable file not found in $PATH: unknown
```

> This is because [Router's image is `FROM scratch`](https://github.com/alphagov/router/blob/9797473edbbcbb5085fdca006bec7f6b1552f4e6/Dockerfile#L7), so bash isn't available.

Author: I'd like to document how you _should_ do the equivalent in Router. There are a few hints in [this Slack thread](https://gds.slack.com/archives/CADKZN519/p1679417676128139) but nothing definitive.

Reviewer: Just exec into the nginx container instead (`-c nginx`), which does have a shell.

Author: Thanks, updated 👍
